### PR TITLE
featuregate UID in RequestHeader authenticator

### LINKED ISSUE #128986: Dueling writes to extension-apiserver-authentication configmap during 1.31 → 1.32.0-rc.1 upgrade
### What happened?

We had an alert firing for high memory usage for 1.32. At the time of increase we changed the churn to create at v1.31 (So update to v1.32). Previously it was creating at v1.30 and updating to v1.31

We can see some etcd are using more memory than others
```
kubx-etcd-02          etcd-csprq8020q6fal7635s0-6zxcz72w5v                         24m          420Mi           
kubx-etcd-02          etcd-csprq8020q6fal7635s0-hm5px7dkc5                         19m          416Mi           
kubx-etcd-02          etcd-csprq8020q6fal7635s0-jhmsq75wc5                         60m          566Mi  

...

kubx-etcd-04          etcd-cssndj620tea3foqjl5g-5qqscgqp2g                         28m          101Mi           
kubx-etcd-04          etcd-cssndj620tea3foqjl5g-c45kz2lprz                         23m          117Mi           
kubx-etcd-04          etcd-cssndj620tea3foqjl5g-wsbwcb9fp8                         50m          103Mi 
```

### What did you expect to happen?

When a cluster updating to 1.32 from 1.31, we expected (roughly) th same etcd memory usage as 1.31

### How can we reproduce it (as minimally and precisely as possible)?

- Deploy a 1.31 cluster
- upgrade to 1.32
- Observer memory usage on etcd

### Anything else we need to know?

A cluster that was created at v1.32 has an etcd `container_memory_working_set_bytes` of around 130 MB.
A v1.31 cluster is also around 130 MB. After the upgrade to v1.32 it increased to around 450 MB (Caused by a short increase in etcd_mvcc_db_total_size_in_use_in_bytes to 430 MB). `etcd_mvcc_db_total_size_in_bytes` permanently increased to 450 MB at upgrade time too. Over time some of the etcds saw an increase in `container_memory_working_set_bytes` to 600MB, but `etcd_mvcc_db_total_size_in_bytes` did not change.

When it was defragged the `etcd_mvcc_db_total_size_in_bytes` dropped to 11 MB (Which is just above the value of `etcd_mvcc_db_total_size_in_use_in_bytes` but `container_memory_working_set_bytes` only dropped to around 450MB.

After further investigation it requires a defrag, followed by restart of etcd pods to return the memory usage to normal after an update to v1.32.

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.31.1
Kustomize Version: v5.4.2
Server Version: v1.32.0-beta.0+IKS
```

</details>

### Cloud provider

<details>
IBM Cloud
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
PRETTY_NAME="Ubuntu 24.04.1 LTS"
NAME="Ubuntu"
VERSION_ID="24.04"
VERSION="24.04.1 LTS (Noble Numbat)"
VERSION_CODENAME=noble
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=noble
LOGO=ubuntu-logo

$ uname -a
Linux test-csug3jq20lvmbeb4nq2g-standard132-default-00000172 6.8.0-48-generic #48-Ubuntu SMP PREEMPT_DYNAMIC Fri Sep 27 14:04:52 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by stlaz**
#### What type of PR is this?
/kind regression
/kind feature

#### What this PR does / why we need it:
Fixes a hotloop during kube-apiserver upgrade.

#### Which issue(s) this PR fixes:

Fixes #128986 

#### Does this PR introduce a user-facing change?

```release-note
Request header UID propagation is gated behind an alpha RemoteRequestHeaderUID feature gate.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/features/kube_features.go [hunk 1]
```diff
@@ -359,6 +366,10 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 		{Version: version.MustParse("1.29"), Default: true, PreRelease: featuregate.GA, LockToDefault: true},
 	},
 
+	RemoteRequestHeaderUID: {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/features/kube_features.go:369]: also register this into `pkg/features/versioned_kube_features.go` like other generic feature gates

File: pkg/controlplane/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go [hunk 2]
```diff
@@ -305,9 +310,12 @@ func getClusterAuthenticationInfoFor(data map[string]string) (ClusterAuthenticat
 	if err != nil {
 		return ClusterAuthenticationInfo{}, err
 	}
-	ret.RequestHeaderUIDHeaders, err = jsonDeserializeStringSlice(data["requestheader-uid-headers"])
-	if err != nil {
-		return ClusterAuthenticationInfo{}, err
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.RemoteRequestHeaderUID) {
+		ret.RequestHeaderUIDHeaders, err = jsonDeserializeStringSlice(data["requestheader-uid-headers"])
```
Comment (liggitt) [inline pkg/controlplane/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go:315]: is it ok to leave `ret.RequestHeaderUIDHeaders` nil if the feature is disabled or do other things assume it is non nil?

**Comment by dims** [comment]
@stlaz we remove a http header here, do we want to add the feature flag around it?

https://github.com/kubernetes/kubernetes/pull/115834/files#diff-ca7a3e3718c8fd64fa4e1ed2b95835a9545d4da18b282a57620c67428ad66b57R129

**Comment by liggitt** [comment]
> @stlaz we remove a http header here, do we want to add the feature flag around it?

I would leave the removal in place ungated. Disabling the feature should result in no uid header propagating from the proxy

**Comment by enj** [comment]
/release-note-edit ```release-note Gate request header UID propagation behind RemoteRequestHeaderUID feature.```

**Comment by enj** [comment]
/release-note-edit ```release-note
Gate request header UID propagation behind RemoteRequestHeaderUID feature.```

**Comment by enj** [comment]
/release-note-edit

```release-note
Gate request header UID propagation behind the RemoteRequestHeaderUID feature.
```

File: test/integration/examples/apiserver_test.go [hunk 3]
```diff
@@ -337,12 +341,12 @@ func testFrontProxyConfig(t *testing.T, withUID bool) {
 		transport.WrapperFunc(func(rt http.RoundTripper) http.RoundTripper {
 			return roundTripperFunc(func(req *http.Request) (*http.Response, error) {
 				gotUser, ok := genericapirequest.UserFrom(req.Context())
-				if !ok || gotUser.GetName() == "system:anonymous" {
```
Comment (liggitt) [inline test/integration/examples/apiserver_test.go:340]: curious about this change, I only expected changes to the uid

**Comment by liggitt** [comment]
changes lgtm, one question about the `system:anonymous` change in the integration test

**Comment by liggitt** [comment]
were you able to verify the behavior with a concurrent 1.31 and 1.32 server (e.g. https://github.com/kubernetes/kubernetes/issues/128986#issuecomment-2516924243)?

File: test/integration/examples/apiserver_test.go [hunk 4]
```diff
@@ -769,7 +779,8 @@ func runPreparedWardleServer(
 		if flunderBanningFeatureGate {
 			args = append(args, "--feature-gates", fmt.Sprintf("wardle:BanFlunder=%v", banFlunder))
 		}
-		wardleCmd := sampleserver.NewCommandStartWardleServer(ctx, wardleOptions)
+		// TODO figure out how to actually make BinaryVersion/EmulationVersion work with Wardle and KAS at the same time when Alpha FG are being set
+		wardleCmd := sampleserver.NewCommandStartWardleServer(ctx, wardleOptions, withUID)
```
Comment (dims) [inline test/integration/examples/apiserver_test.go:783]: the parameter name for new one threw me off somewhat .. we are using `withUID` here and in start.go in the definition of the `NewCommandStartWardleServer` method we are using `skipDefaultComponentGlobalsRegistrySet`

**Comment by liggitt** [comment]
/lgtm
/approve

The off-by-default gate around the change to the configmap publishing / reading _and_ the no-op if the no uid headers are specified will resolve the duel observed in #128986 during upgrade. We can revisit how to avoid dueling when changing from no-value to set-value in 1.33.

/hold
if it's fast to verify with a skewed 1.31 / 1.32-dev setup as in https://github.com/kubernetes/kubernetes/issues/128986#issuecomment-2516924243, please do that. If that will take more than ~30 minutes, go ahead and unhold to get this merged to restore 1.32

**Comment by enj** [comment]
/hold cancel

Confirmed via steps outlined in https://github.com/kubernetes/kubernetes/issues/128986#issuecomment-2516924243 that the regression has been fixed.

**Pull request merged**